File: app/src/main/java/com/example/tanawatk/mvvmandroid/di/component/AppComponent.java
package com.example.tanawatk.mvvmandroid.di.component;

import android.content.Context;

import com.example.tanawatk.mvvmandroid.KnotApp;
import com.example.tanawatk.mvvmandroid.di.module.ApiModule;
import com.example.tanawatk.mvvmandroid.di.module.AppModule;
import com.example.tanawatk.mvvmandroid.di.qualifier.ApplicationContext;
import com.example.tanawatk.mvvmandroid.viewmodel.ViewModelFactory;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(modules = {AppModule.class, ApiModule.class})
public interface AppComponent {

    ViewModelFactory provideViewModelFactory();

    void inject(KnotApp application);
}

File: app/src/main/java/com/example/tanawatk/mvvmandroid/di/component/AppComponent.kt
package com.example.tanawatk.mvvmandroid.di.component

import com.example.tanawatk.mvvmandroid.KnotApp
import com.example.tanawatk.mvvmandroid.di.module.ApiModule
import com.example.tanawatk.mvvmandroid.di.module.AppModule
import com.example.tanawatk.mvvmandroid.di.module.DatabaseModule
import com.example.tanawatk.mvvmandroid.viewmodel.ViewModelFactory
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [AppModule::class, ApiModule::class, DatabaseModule::class])
interface AppComponent {
    fun provideViewModelFactory(): ViewModelFactory
    fun inject(application: KnotApp)
}